src/Native/Runtime/unix/UnixEvent.cpp

```diff
@@ -0,0 +1,334 @@
+
+        m_waiterListEntries = (ListEntry*)malloc(eventCount * sizeof(ListEntry));
+        FATAL_ASSERT(m_waiterListEntries != NULL, "Out of memory allocating m_waiterListEntries array");
+    }
+
+    int st = pthread_mutex_init(&m_mutex, NULL);
+    FATAL_ASSERT(st == 0, "Failed to initialize UnixEventWaiter mutex");
+
+    pthread_condattr_t attrs;
+    st = pthread_condattr_init(&attrs);
+    FATAL_ASSERT(st == 0, "Failed to initialize UnixEventWaiter condition attribute");
+
+#if HAVE_CLOCK_MONOTONIC
+    // Ensure that the pthread_cond_timedwait will use CLOCK_MONOTONIC
+    st = pthread_condattr_setclock(&attrs, CLOCK_MONOTONIC);
+    FATAL_ASSERT(st == 0, "Failed to set UnixEventWaiter condition variable wait clock");
+#endif // HAVE_CLOCK_MONOTONIC
+
+    st = pthread_cond_init(&m_condition, &attrs);
+    FATAL_ASSERT(st == 0, "Failed to initialize UnixEventWaiter condition variable");
+
+    st = pthread_condattr_destroy(&attrs);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter condition attribute");
+}
+
+UnixEventWaiter::~UnixEventWaiter()
+{
+    int st = pthread_mutex_destroy(&m_mutex);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter mutex");
+
+    st = pthread_cond_destroy(&m_condition);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter condition variable");
+
+    if (m_waiterListEntries != m_embeddedWaiterListEntries)
+    {
+        free(m_waiterListEntries);
+    }
+}
+
+// Add event to the waiter
+// Parameters:
+//  index - Index of the event that the waiter reports when the event is signalled
+//  event - Event to wait on
+void UnixEventWaiter::AddEvent(int index, UnixEvent* event)
+{
+    ASSERT(index < m_eventCount);
+    m_waiterListEntries[index] = ListEntry(this);
+    event->AddWaiter(&m_waiterListEntries[index]);
+}
+
+// Remove event from the waiter
+// Parameters:
+//  index - Index of the event to remove
+//  event - Event to remove
+void UnixEventWaiter::RemoveEvent(int index, UnixEvent* event)
+{
+    ASSERT(index < m_eventCount);
+    event->RemoveWaiter(&m_waiterListEntries[index]);
+}
+
+// Signal the waiter that an attached event was set. It is called 
+// by the UnixEvent instances. 
+// Parameters:
+//  index - index of the event in the list of events managed by this waiter
+// Return:
+//  true if the signal has caused the wait to complete, false if the wait
+//  completion was already triggered by another signal.
+bool UnixEventWaiter::Signal(int index)
+{
+	bool releasedWait = false;
+    pthread_mutex_lock(&m_mutex);
+    // Only the first signal is used to release the waiting thread. 
+    // If the wait has already timed out or failed, we don't consume the signal.
+    // This is important for autoreset events
+    if (!m_timedOutOrFailed && (m_eventIndex == NoEventIndex))
+    {
+	    m_eventIndex = index;
+	    releasedWait = true;
+	    // Unblock the thread waiting for the condition variable
+	    pthread_cond_signal(&m_condition);
```

Comment:
Actually, the pthread_cond_signal doc says:
> The pthread_cond_broadcast() or pthread_cond_signal() functions may be called by a thread whether or not it currently owns the mutex that threads calling pthread_cond_wait() or pthread_cond_timedwait() have associated with the condition variable during their waits; however, if predictable scheduling behavior is required, then that mutex shall be locked by the thread calling pthread_cond_broadcast() or pthread_cond_signal().

I've googled a bit and found an article that describes what the less predictable scheduling behavior means:
https://groups.google.com/forum/?hl=ky#!msg/comp.programming.threads/wEUgPq541v8/ZByyyS8acqMJ

So it seems we should be ok with signalling out of the lock as you've suggested.